### ISSUE TITLE
UniqueFactorizationDomains gives wrong judgement

### ISSUE THREAD

**Issue by YijunYuan** [sagemath/sage#36266]
### Steps To Reproduce

```sage
sage: K.<z> = NumberField( x^2+11 )
sage: K.ring_of_integers() in UniqueFactorizationDomains
```

### Expected Behavior

return `True`, as 11 is a [Heegner number](https://en.wikipedia.org/wiki/Heegner_number)

### Actual Behavior

return `False`

### Environment

```markdown
- **OS**: Ubuntu 23.04 lunar(on the Windows Subsystem for Linux)
- **Sage Version**: 10.1 (compiled by myself)
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

**Comment by YijunYuan** [comment]
https://github.com/sagemath/sage/blob/41031292ff1ae518cd5b5a29ce277aa1ff8ced9e/src/sage/categories/unique_factorization_domains.py#L83-L86

return `False` when exception??????
Please at lease give `NotImplementedError`

By the way, a quick search in source code shows that there are only a few (trivial) structures (field, Z/nZ, polynomial ring over UFD,...) support the UFD verification. Therefore this is almost useless.

**Comment by pzq-loves-qwq** [comment]
For rings of integers of number fields, a workaround is to see whether the `class_number()` of the field is 1.

**Comment by YijunYuan** [comment]
> For rings of integers of number fields, a workaround is to see whether the `class_number()` of the field is 1.

You are right. But my point here is that it is a nightmare to return a hard-coded value instead of  `NotImplementedError`.  And this is not the only case of this kind of mess in SageMath. This makes Sage unreliable.

**Comment by YijunYuan** [comment]
> For rings of integers of number fields, a workaround is to see whether the `class_number()` of the field is 1.

And btw, my original need is to test some quotients of a multivariable polynomial ring. I use the Heegner number in the Issue to simplify the MWE.
